refactor(Editing transaction Dates)

lomamech opened the pull request:
- This PR makes it possible to put the period as well as the year whenever
the date changes and prevent the modification of the date for closed or
inexistent fiscal years
- Implement of function lookupFiscalYearByDate for check the FiscalYear
  and period for one Date
closes #2052

lomamech committed:
refactor(Editing transaction Dates)
- This PR makes it possible to put the period as well as the year whenever
the date changes and prevent the modification of the date for closed or
inexistent fiscal years
- Implement of function lookupFiscalYearByDate for check the FiscalYear
  and period for one Date
-

jniles commented on server/controllers/finance/fiscal.js:
```diff
@@ -648,3 +649,25 @@ function getPeriodByFiscal(fiscalYearId) {
+ *
+ * @returns {Promise} - a promise resolving to the fiscal record
+ *
+ * @private
```
This isn't a private method if you expose it :)

jniles commented on server/controllers/finance/fiscal.js:
```diff
@@ -648,3 +649,25 @@ function getPeriodByFiscal(fiscalYearId) {
+ *
+ * @description
+ * This function returns a single record from the fiscal year table matching
+ * the DATE provided.  If no record is found, it throws a NotFound error.
```
I don't think this function throws a `NotFound` error.  If you used `db.one()` it would.

jniles commented on server/controllers/finance/journal/index.js:
```diff
@@ -276,7 +280,28 @@ function editTransaction(req, res, next) {
+          throw new BadRequest('Closed Fiscal Year', 'POSTING_JOURNAL.ERRORS.CLOSED_FISCAL_YEAR');
+        }
 
+        _.each(rowsChanged, (row) => {
```
This function alters data in the transaction.   @sfount attempted to avoid [spaghetti code](https://en.wikipedia.org/wiki/Spaghetti_code) by putting all logic that changes the transaction in the function `transformColumns()`.   This code belongs there.

jniles commented on server/controllers/finance/journal/index.js:
```diff
@@ -452,6 +477,17 @@ function transformColumns(rows, newRecord, oldTransaction) {
 }
 
+//Check the Valid Update of Date
+function selectTransDate(rows){
```
I don't think this function does any validation.  If anything, it should just do this:
```js
const selectTransDate = (rows) => rows[0].trans_date;
```

jniles reviewed (changes requested):
@lomamech, while this accomplishes the original goal, it should really conform to the expectations of the module.

Any code that alters or _transforms_ the transaction should appear in the function `transformColumns` or be called from `transformColumns`.  That makes it much easier to refactor/maintain in the future.

lomamech committed:
refactor(Editing transaction Date)
- Fix the comment for the description of function
- Put all alteration of data in the function transformColumns()
- rename the name of function selectTransdate
